### PR TITLE
[onert] Change memory planners to class template

### PR DESCRIPTION
**Pull request by ragmani**
This commit changes memory planners to class template to allow memory planning for variable indexes.
  - Change `IMemoryPlanner` to template
  - Change memory planners to class template
  - Change the factory method of memory planners to template

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

---

For #12564

### CODE REVIEW

**Review by jyoungyun — APPROVED** [review]
LGTM

File: runtime/onert/core/src/backend/basic/MemoryPlanner.h [hunk 1]
```diff
@@ -64,30 +65,45 @@ class BumpPlanner : public IMemoryPlanner
...
+
+  VERBOSE(BP_PLANNER) << "CLAIM(" << ind << "): " << blk.offset << ", " << blk.size << std::endl;
+}
+
+template <typename Index> inline void BumpPlanner<Index>::release(const Index &ind)
+{
+  VERBOSE(BP_PLANNER) << "RELEASE(" << ind << "): "
+                      << "NOTHING does" << std::endl;
+}
+
```
Comment (ragmani) [inline runtime/onert/core/src/backend/basic/MemoryPlanner.h:89]: From the removed `MemoryManager.cc`

File: runtime/onert/core/src/backend/basic/MemoryPlanner.h [hunk 2]
```diff
@@ -97,34 +113,100 @@ class FirstFitPlanner : public IMemoryPlanner
...
+      _claim_table.erase(it);
+
+      VERBOSE(FF_PLANNER) << "release(" << ind << "): [+" << offset << ", " << size << "sz]"
+                          << std::endl;
+      return;
+    }
+  }
+  assert(!"Cannot release for given index. It has been not claimed or released already.");
+}
+
```
Comment (ragmani) [inline runtime/onert/core/src/backend/basic/MemoryPlanner.h:190]: From the removed `MemoryManager.cc`

File: runtime/onert/core/src/backend/basic/MemoryPlanner.h [hunk 3]
```diff
@@ -139,20 +221,119 @@ class WICPlanner : public IMemoryPlanner
...
+  _indices.clear();
+}
+
+template <typename Index> inline std::unordered_map<Index, Block> &WICPlanner<Index>::memory_plans()
+{
+  if (!_initialized)
+    buildMemoryPlans();
+  return _mem_plans;
+}
+
```
Comment (ragmani) [inline runtime/onert/core/src/backend/basic/MemoryPlanner.h:336]: From the removed `MemoryManager.cc`

**Review by hseok-oh — APPROVED** [review]
LGTM

**Pull request merged**